test/unit/ui_utils_spec.js

```diff
@@ -398,4 +399,236 @@ describe('ui_utils', function() {
       }, done.fail);
     });
   });
+
+  describe('getVisibleElements', function() {
+    // These values are based on margin/border values in the CSS, but there
+    // isn't any real need for them to be; they just need to take *some* value.
+    const BORDER_WIDTH = 9;
+    const SPACING = 2 * BORDER_WIDTH - 7;
+
+    // This is a helper function for assembling an array of view stubs from an
+    // array of arrays of [width, height] pairs, which represents wrapped lines
+    // of pages. It uses the above constants to add realistic spacing between
+    // the pages and the lines.
+    function makePages(lines) {
+      const result = [];
+      let lineTop = 0, id = 0;
+      for (const line of lines) {
+        const lineHeight = line.reduce(function(h, pair) {
+          return Math.max(h, pair[1]);
+        }, 0);
+        let offsetLeft = 0;
+        for (const [clientWidth, clientHeight] of line) {
+          const offsetTop =
+            lineTop + (lineHeight - clientHeight) / 2 - BORDER_WIDTH;
```

Comment:
You are technically correct, but again, this detail would add unnecessary complexity to these tests. It's enough of a chore to look at those matrixes of page dimensions and puzzle out the geometry from them; thinking about the full CSS box model while doing so would be pure make-work for the reader, for no benefit.

Reply:
> For example, if you ran one of these tests with `makePages([[[20, 50]]])` and a `scrollTop` of 50, you should expect the page to be hidden, from an intuitive reading of those numbers 

Actually, I expect the page to be visible, because there is a `BORDER_WIDTH` above the page, so `BORDER_WIDTH` pixels of the page's bottom are sticking out at the top of the viewport.

I think that there is a disconnect in assumptions that we make (and from a quick flance don't fully understand where you are going with your last reply), so I will wait for your comments before reviewing the specific calculations from this comment in detail.

> > and assume that `margin-left:0`
>
> I don't know what you mean by that?

Ehr, I meant `margin-top:0`. If you assume that the page has `margin-top:0`, then there is no need to subtract a margin, *depending on how you view the distribution of space*.

To me, `SPACING = margin top + border top + border bottom + margin bottom`.
If you assume that `margin-top` is `0`, then the page's top border can start at offset 0 and you don't need to subtract anything (since `offsetTop` is the position relative to the point right above the top border). Otherwise `offsetTop` needs to start at the position given by the value of `margin-top` (in all cases, assuming that the container (`.pdfViewer` in reality) has no extra margins).